src/meta/processors/admin/Balancer.cpp

```diff
@@ -0,0 +1,264 @@
+            running_ = false;
+            return;
+        }
+        while (iter->valid()) {
+            auto spaceId = MetaServiceUtils::spaceId(iter->key());
+            spaces.push_back(spaceId);
+            iter->next();
+        }
+    }
+    plan_ = std::make_unique<BalancePlan>(time::TimeUtils::nowInSeconds(), kv_);
+    for (auto spaceId : spaces) {
+        auto tasks = genTasks(spaceId);
+        for (auto& task : tasks) {
+            plan_->addTask(std::move(task));
+        }
+    }
+    plan_->onFinished_ = [this] () {
+        plan_.reset();
+        bool expected = true;
+        CHECK(running_.compare_exchange_strong(expected, false));
+    };
+    plan_->registerTaskCb();
+    if (!plan_->saveInStore()) {
+        LOG(INFO) << "Can't persist the plan!";
+        plan_->onFinished_();
+    }
+}
+
+std::vector<BalanceTask> Balancer::genTasks(GraphSpaceID spaceId) {
+    std::unordered_map<HostAddr, std::vector<PartitionID>> hostParts;
+    int32_t totalParts = 0;
+    getHostParts(spaceId, hostParts, totalParts);
+    if (totalParts == 0 || hostParts.empty()) {
+        LOG(ERROR) << "Invalid space " << spaceId;
+        return std::vector<BalanceTask>();
+    }
+    auto activeHosts = ActiveHostsMan::instance()->getActiveHosts();
+    std::vector<HostAddr> newlyAdded;
+    std::vector<HostAddr> lost;
+    calDiff(hostParts, activeHosts, newlyAdded, lost);
+    decltype(hostParts) newHostParts(hostParts);
+    for (auto& h : newlyAdded) {
+        newHostParts.emplace(h, std::vector<PartitionID>());
+    }
+    for (auto& h : lost) {
+        newHostParts.erase(h);
+    }
+    LOG(INFO) << "Now, try to balance the newHostParts";
+    // We have two parts need to balance, the first one is parts on lost hosts
+    // The seconds one is parts on unbalanced host in newHostParts.
+    std::vector<BalanceTask> tasks;
+    for (auto& h : lost) {
+        auto& lostParts = hostParts[h];
+        for (auto& partId : lostParts) {
+            auto ret = pickupHost(newHostParts, partId);
+            if (!ret.ok()) {
+                LOG(ERROR) << "Error:" << ret.status();
+                return std::vector<BalanceTask>();
+            }
+            auto& luckyHost = ret.value();
+            newHostParts[luckyHost].emplace_back(partId);
+            tasks.emplace_back(partId, h, luckyHost, kv_);
+        }
+    }
+    if (newHostParts.size() < 2) {
+        LOG(INFO) << "Too few hosts, no need for balance!";
+        return tasks;
+    }
+    balanceParts(newHostParts, totalParts, tasks);
+    return tasks;
+}
+
+void Balancer::balanceParts(std::unordered_map<HostAddr, std::vector<PartitionID>>& newHostParts,
+                            int32_t totalParts,
+                            std::vector<BalanceTask>& tasks) {
+    float avgLoad = static_cast<float>(totalParts)/newHostParts.size();
+    LOG(INFO) << "The expect avg load is " << avgLoad;
+    int32_t minLoad = avgLoad * 0.9;
```

Comment:
Does this criteria is fixed or tunable?  It would be better to make it tunable(a config item)? The smaller the gap, the bigger the rebalance load

Reply:
I have updated the minLoad equals floor(avgLoad) and maxLoad equals ceil(avgLoad) now.